src/frequenz/sdk/util/ringbuffer.py

```diff
@@ -0,0 +1,259 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Ringbuffer implementation with focus on time & memory efficiency."""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from datetime import datetime, timedelta
+from typing import Any, Generic, Sequence, TypeVar, Union
+
+import numpy as np
+
+T = TypeVar("T")
+
+Container = Union[list, np.ndarray]
+
+
+class OrderedRingBuffer(Generic[T]):
+    """Time aware ringbuffer that keeps its entries sorted by time."""
+
+    def __init__(
+        self,
+        buffer: Any,
+        sample_rate: float,
+        window_border: datetime = datetime(1, 1, 1),
+    ) -> None:
+        """Initialize the time aware ringbuffer.
+
+        Args:
+            buffer: instance of a buffer container to use internally
+            sample_rate: resolution of the incoming timestamps in
+                hz (samples per second)
+            window_border: datetime depicting point in time to use as border
+                beginning, useful to make data start at the beginning of the day or
+                hour.
+        """
+        self._buffer = buffer
+        self._sample_rate = sample_rate
+        self._window_start = window_border
+
+        self._missing_windows = []
+        self._datetime_newest = datetime.min
+        self._datetime_oldest = datetime.max
+
+    @property
+    def missing_windows(self) -> list:
+        """Get the list of missing windows.
+
+        Returns:
+            list of missing windows.
+        """
+        return self._missing_windows
+
+    @property
+    def maxlen(self) -> int:
+        """Get the max length.
+
+        Returns:
+            the max amount of items this container can hold.
+        """
+        return len(self._buffer)
+
+    def update(self, timestamp: datetime, value: T, missing: bool = False) -> bool:
```

Comment:
What does it mean that `value` is missing?  There is no test with `missing = True` .
Should it be None? Or it should has value?

Reply:
I def. should add a test for `missing = True`, totally missed that.

Value can be anything you want, it will still be written but it's marked as "missing" in our `missing_windows()`.

The idea was to allow max. flexibility. Some users want to write Nan values, some want to use the avg of the previous one and for some the datatype might not even allow an easy "invalid" value (e.g. int values).

With this method it's independent of the value written.

Reply:
but how would you specify what should be used?

Reply:
What do you mean? I am not sure what your "what" refers to and what that "what" would be used for?